lib/Map.fram

```diff
@@ -0,0 +1,502 @@
+  | (Exclusion left, Inclusion right) =>
+      let (_,middle) = partionLt compare tree left in 
+      let (result,_) = partionLt compare middle right in result
+  | (Exclusion left, Exclusion right) =>
+      let (_,middle) = partionLt compare tree left in
+      let (result,_) = partionGt compare middle right in result
+  end
+
+data MapT Key Val = MapT of Tree (Pair Key Val)
+
+// Wrappers
+let isEmptyT (MapT tree) = isEmpty tree
+
+let insertT compare (MapT tree) key val  = MapT (insert compare tree key val)
+
+let insertChangeT compare (MapT tree) key val = 
+  let (tree,bool) = insert' compare tree key val in (MapT tree, bool)
+
+let removeT compare (MapT tree) key = MapT (remove compare tree key)
+
+let removeChangeT compare (MapT tree) key = 
+  let (tree,bool) = remove' compare tree key in (MapT tree, bool)
+
+let memberT compare (MapT tree) key = member compare tree key
+
+let findT compare (MapT tree) key = find compare tree key
+
+let operateT compare (MapT tree) key absentf presentf =
+  let (val,res,tree) = operate compare tree key absentf presentf 
+    in (val,res, MapT tree)
+
+let mapFoldl (MapT tree) func acc = foldl func tree acc
+
+let mapFoldr (MapT tree) func acc = foldr func tree acc
+
+let toListT (MapT tree) = toList tree
+
+let toValueListT (MapT tree) = toValueList tree
+
+let domainT (MapT tree) =  domain tree
+
+let mapVal (MapT tree) func = MapT (map tree func)
+
+let mapKey (MapT tree) func = MapT (map2 tree func)
+
+let appT (MapT tree) func = app tree func
+
+let unionT compare (MapT tree1) (MapT tree2) merge = 
+  MapT (union compare tree1 tree2 merge)
+
+let partionT compare (MapT tree) key = 
+  let (t1,v,t2) = partion compare tree key 
+    in (MapT t1, v, MapT t2)
+
+let partionLtT compare (MapT tree) key = 
+  let (t1,t2) = partionLt compare tree key
+    in (MapT t1, MapT t2)
+
+let partionGtT compare (MapT tree) key = 
+  let (t1,t2) = partionGt compare tree key
+    in (MapT t1, MapT t2)
+
+let rangeT compare (MapT tree) left right = 
+  MapT (range compare tree left right)
+
+let lowerBoundT (MapT tree) = lowerBound tree
+
+let lowerBoundTErr (MapT tree) = lowerBoundErr tree
+
+let upperBoundT (MapT tree) = upperBound tree
+
+let upperBoundTErr (MapT tree) = upperBoundErr tree
+
+let lowerBoundGtT compare (MapT tree) key = lowerBoundGt compare tree key
+
+let lowerBoundGtTErr compare (MapT tree) key =
+  lowerBoundGtErr compare tree key
+
+let upperBoundLtT compare (MapT tree) key = upperBoundLt compare tree key
+
+let upperBoundLtTErr compare (MapT tree) key =
+  upperBoundLtErr compare tree key
+
+let lowerBoundGeqT compare (MapT tree) key = lowerBoundGeq compare tree key
+
+let lowerBoundGeqTErr compare (MapT tree) key =
+  lowerBoundGeqErr compare tree key
+
+let upperBoundLeqT compare (MapT tree) key = upperBoundLeq compare tree key
+
+let upperBoundLeqTErr compare (MapT tree) key = 
+  upperBoundLeqErr compare tree key
+
+pub let make {Key} (compare : Key -> Key -> [] Ordered) = Map {
```

Comment:
Can you add another `make` (I don't know what the name should be), that will take as an argument type of key, and get comparator from method compare?
The same would be useful with sets